source/hpack/decoder.d

```diff
@@ -0,0 +1,278 @@
+module HPACK.decoder;
+
+import HPACK.exception;
+import HPACK.huffman;
+import HPACK.tables;
+import HPACK.util;
+
+import std.range; // Decoder
+import std.bitmanip; // prefix encoding / decoding
+
+/** Module to implement an header decoder consistent with HPACK specifications (RFC 7541)
+  * The detailed description of the decoding process, examples and binary format details can
+  * be found at:
+  * Section 3: https://tools.ietf.org/html/rfc7541#section-3
+  * Section 6: https://tools.ietf.org/html/rfc7541#section-6
+  * Appendix C: https://tools.ietf.org/html/rfc7541#appendix-C
+*/
+alias HTTP2SettingValue = uint;
+
+/** implements an input range to decode an header block
+  * m_table is a reference to the original table
+  */
+struct HeaderDecoder(T = ubyte[])
+		if (isInputRange!T && (is(ElementType!T : char) || (is(ElementType!T : ubyte))))
+{
+	private {
+		ubyte[] m_range;
+		IndexingTable m_table; // only for retrieving data
+		HTTP2HeaderTableField[] m_decoded;
```

Comment:
This currently requires storing the decoded header in GC memory. It would be nice to instead either have a lazily decoded input range interface, where only the currnt header field is ever stored, or (a bit less flexible, but still okay), invoke a callback of some form with a `scope HTTP2HeaderTableField` each time a field has been read.

Storing all fields as an array can then be easily layered on top, but at the same time it would be made possible to write a request processor that would be `@nogc` (unless I'm missing another allocation location right now) and greatly improves performance, especially in multi-threaded operation (the central GC lock is pretty bad when it comes to multi-threading performance, as well as triggering collection runs).